### PR TITLE
Fix check_backward

### LINKED ISSUE #3541: Bad gradients pass gradient check
Even when the computed gradients are wrong, the gradient check tests seem to pass.
This started occurring since merge https://github.com/chainer/chainer/commit/85372a2d909f1649a83aab4fd3c4fab0dfa6e732

It is easily reproducible, by simply altering the outputs of some function and then running the corresponding test, noticing that the test still passes.

### PR DESCRIPTION
**Pull request by niboshi**
Fixes #3541

### CODE REVIEW

File: chainer/gradient_check.py [hunk 1]
```diff
@@ -282,7 +282,7 @@ def check_backward(func, x_data, y_grad, params=(),
                 raise RuntimeError(
                     'gradients of some arguments are not calculated')
 
-    if len(xs) - len(no_grads) + len(params) == 0:
+    if len(xs) - sum(1 for _ in no_grads if _) + len(params) == 0:
```
Comment (unnonouno) [inline chainer/gradient_check.py]: `no_grads.count(True)` ?

File: tests/chainer_tests/test_gradient_check.py [hunk 2]
```diff
@@ -370,6 +370,55 @@ def f(x, y):
         gradient_check.check_backward(f, (x1, x2), g1, no_grads=[False, True])
 
 
+class TestCheckBackwardFailure(unittest.TestCase):
+
+    def _broken_func_1(self):
```
Comment (unnonouno) [inline tests/chainer_tests/test_gradient_check.py]: Why did you wrap it with a function?

File: tests/chainer_tests/test_gradient_check.py [hunk 3]
```diff
@@ -370,6 +370,94 @@ def f(x, y):
...
+
+        return Broken()
+
+    def _broken_func_2(self):
+        class Broken(chainer.FunctionNode):
+            def forward(self, inputs):
+                x, = inputs
+                return (x * x),
+
+            def backward(self, inputs, grad_outputs):
```
Comment (unnonouno) [inline tests/chainer_tests/test_gradient_check.py]: `(self, indexes, grad_outputs)` ?

Comment (niboshi) [reply tests/chainer_tests/test_gradient_check.py]: Fixed.

File: chainer/gradient_check.py (hunk 1 above)
Comment (niboshi) [reply chainer/gradient_check.py]: Fixed.

File: tests/chainer_tests/test_gradient_check.py (hunk 2 above)
Comment (niboshi) [reply tests/chainer_tests/test_gradient_check.py]: Because they could be reused.
Indeed, `_broken_func_3` is used twice.

Comment (unnonouno) [reply tests/chainer_tests/test_gradient_check.py]: ok

**Comment by niboshi** [comment]
Hmm, it seems it's numerically unstable than before.

**Comment by unnonouno** [comment]
I think so. In the previous implementation, the method calculates gradient at the origin point. But this implementation calculates gradient at a random position.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**